### PR TITLE
 gltfio: support specular-glossiness

### LINKED ISSUE #1068: Support for KHR_materials_pbrSpecularGlossiness in glTF I/O
I'm not actually using the GLTF importer that @prideout has written, but it's a very valuable reference for translating the GLTF spec to the Filament engine in my own importer. Are then any plans to support the [KHR_materials_pbrSpecularGlossiness](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_pbrSpecularGlossiness) extension?

I'm guessing that _diffuse_ would map to _baseColor_ and _glossiness_ would map to _1 - roughness_, but is there a suitable mapping for _specular_?

### PR DESCRIPTION
**Pull request by prideout**
Fixes #1068

### CODE REVIEW

File: libs/gltfio/materials/ubershader.mat.in [hunk 1]
```diff
@@ -75,13 +80,48 @@ fragment {
...
+                    glossiness *= sg.a;
+                    vec3 specColor = sg.rgb * materialParams.specularFactor;
+
+                    // NOTE: this conversion logic is a carry-over from MeshAssimp. According to the
+                    // spec: "The specular property from specular-glossiness material model is the
+                    // same as the base color value from the metallic-roughness material model for
+                    // metals."
+                    material.reflectance = 0.5;
+                    material.metallic = 0.0;
+                    if (specColor.r != specColor.g && specColor.r != specColor.b) {
```
Comment (romainguy) [inline libs/gltfio/materials/ubershader.mat.in]: That would merit comparing against the spec or revisiting. I'm not sure how correct this is

File: libs/gltfio/src/MaterialGenerator.cpp [hunk 2]
```diff
@@ -126,10 +129,43 @@ static std::string shaderFromKey(const MaterialKey& config) {
...
+                shader += R"SHADER(
+                    vec4 sg = texture(materialParams_metallicRoughnessMap, metallicRoughnessUV);
+                    glossiness *= sg.a;
+                    vec3 specColor = sg.rgb * materialParams.specularFactor;
+
+                    // NOTE: this conversion logic is a carry-over from MeshAssimp. According to the
+                    // spec: "The specular property from specular-glossiness material model is the
+                    // same as the base color value from the metallic-roughness material model for
+                    // metals."
+                    material.reflectance = 0.5;
```
Comment (romainguy) [inline libs/gltfio/src/MaterialGenerator.cpp]: We shouldn't set reflectance. 0.5 is the default.

File: libs/gltfio/materials/ubershader.mat.in (hunk 1 above)
Comment (prideout) [reply libs/gltfio/materials/ubershader.mat.in]: For starters I think we should simplify this by removing the weird conditionals and just say: "material.baseColor.rgb = specColor" unconditionally.  Many models seem look to reasonable when I tried this.  Need to think more about handling their concept of "diffuse color" correctly though, that's tricky...

File: libs/gltfio/src/MaterialGenerator.cpp (hunk 2 above)
Comment (prideout) [reply libs/gltfio/src/MaterialGenerator.cpp]: Will do.

File: libs/gltfio/materials/ubershader.mat.in (hunk 1 above)
Comment (romainguy) [reply libs/gltfio/materials/ubershader.mat.in]: Doesn't it mean that everything using the specularGlossiness model would end up being a metal? I imagine a plastic in that model has a white specular RGB color and a colored diffuse color.

I thought about this some more and as much as I hate it, the only proper solution is to add a new material model (and we should discourage its usage). This model would have `baseColor` for `diffuse` and add a specularColor (or reuse sheenColor?) attribute. `metallic` and `reflectance` would go away.

**Pull request closed without merging**